### DISCUSSION TITLE
[FEAT]How to resume a workflow instance when it failed？

### DISCUSSION THREAD

**Discussion by coolshun** [elsa-workflows/elsa-core#6215] [Q&A]
When the workflow instance failed ,I want 't to resume(retry)it ,but I couldn't find the api endpoint.What should I do?

**Comment by sfmskywalker** [comment]
When a workflow instance has faulted, it is not intended to be retried. However, if you want to allow your workflow instance to remain in the Running state despite the occurrence of activity faults, you can do so by configuring the [Incident Strategy](https://v3.elsaworkflows.io/docs/incidents/introduction). In your case, you'll want to set it to `ContinueWithIncidentsStrategy`. Contrast to the default `FaultStrategy`, this strategy will not fault the workflow instance.

Then in order to retry the workflow instance, you can use the [Alterations API](https://v3.elsaworkflows.io/docs/alterations/introduction). An alteration is literally an alteration on a given (set of) workflows that you wish to apply. One such alteration is the `ScheduleActivity`, which will reschedule a given activity and then execute the workflow.

**Reply by amTrinh** [reply]
Hi @sfmskywalker , How to retry only the failed activity (including the activities following it) in Elsa V3.2?

**Reply by coolshun** [reply]
please custom a Activity,make a bookmark when it failed,and then trigger the bookmark when you want to retryit .

**Reply by amTrinh** [reply]
> please custom a Activity,make a bookmark when it failed,and then trigger the bookmark when you want to retryit .
@coolshun 
Can i retain the state of an activity and variables of a workflow by using a bookmark?

**Reply by coolshun** [reply]
Yes,please  refer to "RunTask" or "Event" Activity source code.

**Comment by shouha85** [comment]
i have the same problem here , i need to retry the faulted workflow after i have fixthe issue like when exception is raised from custom activity of builtin one then i need to retry the workflow instance run as i don't to start new instance and force the whole set of approvals to re approve 

if i change the incident strategy then the workflow instance will not appear as faulted as the status will be Running and the substatus will be Suspended

and for creating a custom activity and make a bookmark this lead us to rewrite the whole activies again and we will stuck with workflow instance status as it will not be faulted

**Comment by sfmskywalker** [comment]
Current Elsa releases include a retry endpoint in the Alterations module. It builds `ScheduleActivity` alterations for the selected activity IDs, runs them for the workflow instance, and then dispatches the instance. The `ScheduleActivity` handler explicitly transitions an existing faulted activity back to `Running` before scheduling it.

That means you no longer need to redesign every activity around a bookmark simply to retry a faulted activity. For a controlled retry, target the specific failed activity (and choose any following activities deliberately) through the workflow retry/alteration API. Because retrying can repeat side effects, make the activity idempotent or add an appropriate compensating/guard strategy.

The earlier guidance about `ContinueWithIncidentsStrategy` still applies when you want to prevent a workflow from entering the faulted substatus in the first place, but it is not required solely because you need the current retry capability.

Relevant source: `src/modules/Elsa.Alterations/Endpoints/Workflows/Retry/Endpoint.cs`, `src/modules/Elsa.Alterations/AlterationHandlers/ScheduleActivityHandler.cs`, and `src/modules/Elsa.Workflows.Core/IncidentStrategies/FaultStrategy.cs`.
